dsgrid/units/energy.py

```diff
@@ -30,93 +71,237 @@ def to_kwh(unit_col, value_col):
+        F.when(F.col(unit_col) == KWH, (F.col(value_col) * KILO_TO_TERA))
+        .when(F.col(unit_col) == MWH, (F.col(value_col) * MEGA_TO_TERA))
+        .when(F.col(unit_col) == GWH, (F.col(value_col) * GIGA_TO_TERA))
         .when(F.col(unit_col) == TWH, F.col(value_col))
-        .when(F.col(unit_col) == THERM, (F.col(value_col) * KWH_PER_THERM / 1_000_000_000))
-        .when(F.col(unit_col) == MBTU, (F.col(value_col) * KWH_PER_MBTU / 1_000_000_000))
+        .when(F.col(unit_col) == THERM, (F.col(value_col) * THERM_TO_TWH))
+        .when(F.col(unit_col) == MBTU, (F.col(value_col) * MBTU_TO_TWH))
+        .when(F.col(unit_col) == "", F.col(value_col))
         .otherwise(None)
     )
 
 
 def to_therm(unit_col, value_col):
     """Convert a column to therm."""
     return (
-        F.when(F.col(unit_col) == KWH, (F.col(value_col) / KWH_PER_THERM))
-        .when(F.col(unit_col) == MWH, (F.col(value_col) / KWH_PER_THERM / 1_000))
-        .when(F.col(unit_col) == GWH, (F.col(value_col) / KWH_PER_THERM / 1_000_000))
-        .when(F.col(unit_col) == TWH, (F.col(value_col) / KWH_PER_THERM / 1_000_000_000))
+        F.when(F.col(unit_col) == KWH, (F.col(value_col) * KWH_TO_THERM))
+        .when(F.col(unit_col) == MWH, (F.col(value_col) * MWH_TO_THERM))
+        .when(F.col(unit_col) == GWH, (F.col(value_col) * GWH_TO_THERM))
+        .when(F.col(unit_col) == TWH, (F.col(value_col) * TWH_TO_THERM))
         .when(F.col(unit_col) == THERM, F.col(value_col))
-        .when(F.col(unit_col) == MBTU, (F.col(value_col) * THERMS_PER_MBTU))
+        .when(F.col(unit_col) == MBTU, (F.col(value_col) * MBTU_TO_THERM))
+        .when(F.col(unit_col) == "", F.col(value_col))
         .otherwise(None)
     )
 
 
 def to_mbtu(unit_col, value_col):
     """Convert a column to MBtu."""
     return (
-        F.when(F.col(unit_col) == KWH, (F.col(value_col) / KWH_PER_MBTU))
-        .when(F.col(unit_col) == MWH, (F.col(value_col) / KWH_PER_MBTU / 1_000))
-        .when(F.col(unit_col) == GWH, (F.col(value_col) / KWH_PER_MBTU / 1_000_000))
-        .when(F.col(unit_col) == TWH, (F.col(value_col) / KWH_PER_MBTU / 1_000_000_000))
-        .when(F.col(unit_col) == THERM, (F.col(value_col) / THERMS_PER_MBTU))
+        F.when(F.col(unit_col) == KWH, (F.col(value_col) * KWH_TO_MBTU))
+        .when(F.col(unit_col) == MWH, (F.col(value_col) * MWH_TO_MBTU))
+        .when(F.col(unit_col) == GWH, (F.col(value_col) * GWH_TO_MBTU))
+        .when(F.col(unit_col) == TWH, (F.col(value_col) * TWH_TO_MBTU))
+        .when(F.col(unit_col) == THERM, (F.col(value_col) * THERM_TO_MBTU))
         .when(F.col(unit_col) == MBTU, F.col(value_col))
+        .when(F.col(unit_col) == "", F.col(value_col))
         .otherwise(None)
     )
 
 
-def convert_units(df, columns, from_records, from_to_records, to_unit_records) -> DataFrame:
+def from_any_to_any(from_unit_col, to_unit_col, value_col):
+    """Convert a column based on from/to columns."""
+    return (
```

Comment:
Can we daisy-chain the to_XXX() we already created here or do a match case?
```
 if to_unit_col == MWH:
   return to_mbtu(from_unit_col, value_col)
 ```

Reply:
Nice suggestion. I am very surprised that this worked.

Reply:
Cool! Wasn't sure that daisy-training would work either.